pkg/util/worker_test.go

```diff
@@ -0,0 +1,174 @@
+package util
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"testing"
+	"time"
+
+	"go.uber.org/atomic"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+)
+
+func newTestAsyncWorker(reconcileFunc ReconcileFunc) *asyncWorker {
+	options := Options{
+		Name:          "test_async_worker",
+		KeyFunc:       MetaNamespaceKeyFunc,
+		ReconcileFunc: reconcileFunc,
+		RateLimiterOptions: ratelimiterflag.Options{
+			RateLimiterBaseDelay:  time.Millisecond,
+			RateLimiterMaxDelay:   10 * time.Millisecond,
+			RateLimiterQPS:        5000,
+			RateLimiterBucketSize: 100,
+		},
+	}
+	worker := NewAsyncWorker(options)
+
+	return worker.(*asyncWorker)
+}
+
+func Test_asyncWorker_Enqueue(t *testing.T) {
+	const name = "fake_node"
+
+	worker := newTestAsyncWorker(nil)
+
+	node := &corev1.Node{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+	}
+
+	worker.Enqueue(node)
+
+	item, _ := worker.queue.Get()
+
+	if name != item {
+		t.Errorf("Added Item: %v, want: %v", item, name)
+	}
+}
+
+func Test_asyncWorker_AddAfter(t *testing.T) {
+	const name = "fake_node"
+	const duration = 1 * time.Second
+
+	worker := newTestAsyncWorker(nil)
+
+	start := time.Now()
+	worker.AddAfter(name, duration)
+
+	item, _ := worker.queue.Get()
+	end := time.Now()
+
+	if name != item {
+		t.Errorf("Added Item: %v, want: %v", item, name)
+	}
+
+	elapsed := end.Sub(start)
+	if elapsed < duration {
+		t.Errorf("Added Item should dequeued after %v, but the actually elapsed time is %v.",
+			duration.String(), elapsed.String())
+	}
+}
+
+type asyncWorkerReconciler struct {
+	lock sync.Mutex
+	// guarded by lock
+	processed []int
+	// guarded by lock
+	shouldPass map[int]struct{}
+}
+
+func newAsyncWorkerReconciler() *asyncWorkerReconciler {
+	return &asyncWorkerReconciler{
+		shouldPass: make(map[int]struct{}),
+	}
+}
+
+func (a *asyncWorkerReconciler) reconcileFunc(key QueueKey) error {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	v := key.(int)
+	if _, ok := a.shouldPass[v]; !ok {
+		// every item retry once
+		a.shouldPass[v] = struct{}{}
+		return errors.New("fail once")
+	}
+
+	a.processed = append(a.processed, v)
+	return nil
+}
+
+func (a *asyncWorkerReconciler) getProcessed() []int {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	dst := make([]int, len(a.processed))
+	copy(dst, a.processed)
+	return dst
+}
+
+func Test_asyncWorker_Run(t *testing.T) {
+	const cnt = 2000
+
+	reconcile := newAsyncWorkerReconciler()
+	worker := newTestAsyncWorker(reconcile.reconcileFunc)
+	stopChan := make(chan struct{})
+	worker.Run(5, stopChan)
+
+	for i := 0; i < cnt; i++ {
+		worker.Add(i)
+	}
+
+	time.Sleep(time.Second)
+
+	close(stopChan)
```

Comment:
defer close(stopChan)

Reply:
I intend to stop the worker after 1s. defer will close it on method exit.